core/server/master/src/main/java/alluxio/master/file/UfsSyncChecker.java

```diff
@@ -41,65 +41,100 @@
 public final class UfsSyncChecker {
 
   /** UFS directories for which list was called. */
-  private Map<String, UnderFileStatus[]> mListedDirectories;
+  private Map<String, UfsStatus[]> mListedDirectories;
 
   /** This manages the file system mount points. */
   private final MountTable mMountTable;
 
+  /** Directories in sync with the UFS. */
+  private Map<AlluxioURI, Inode<?>> mSyncedDirectories = new HashMap<>();
+
+  /** Whether to skip sync check. */
+  private final boolean mUnchecked;
+
   /**
    * Create a new instance of {@link UfsSyncChecker}.
    *
    * @param mountTable to resolve path in under storage
+   * @param unchecked whether to skip sync check
    */
-  public UfsSyncChecker(MountTable mountTable) {
+  public UfsSyncChecker(MountTable mountTable, boolean unchecked) {
     mListedDirectories = new HashMap<>();
     mMountTable = mountTable;
+    mUnchecked = unchecked;
   }
   /**
    * Check if immediate children of directory are in sync with UFS.
    *
    * @param inode read-locked directory to check
    * @param alluxioUri path of directory to to check
-   * @return true is contents of directory match
    */
-  public boolean isUFSDeleteSafe(InodeDirectory inode, AlluxioURI alluxioUri)
+  public void checkDirectory(InodeDirectory inode, AlluxioURI alluxioUri)
       throws FileDoesNotExistException, InvalidPathException, IOException {
     Preconditions.checkArgument(inode.isPersisted());
 
-    UnderFileStatus[] ufsChildren = getChildrenInUFS(alluxioUri);
-    // Empty directories are not persisted
-    if (ufsChildren == null) {
-      return true;
-    }
-    // Sort-merge compare
-    Arrays.sort(ufsChildren, new Comparator<UnderFileStatus>() {
-      @Override
-      public int compare(UnderFileStatus a, UnderFileStatus b) {
-        return a.getName().compareTo(b.getName());
-      }
-    });
-    int numInodeChildren = inode.getChildren().size();
-    Inode[] inodeChildren = new Inode[numInodeChildren];
-    inodeChildren = inode.getChildren().toArray(inodeChildren);
-    Arrays.sort(inodeChildren, new Comparator<Inode>() {
-      @Override
-      public int compare(Inode a, Inode b) {
-        return a.getName().compareTo(b.getName());
-      }
-    });
-    int ufsPos = 0;
-    int inodePos = 0;
-    while (ufsPos < ufsChildren.length && inodePos < numInodeChildren) {
-      String ufsName = ufsChildren[ufsPos].getName();
-      if (ufsName.endsWith(AlluxioURI.SEPARATOR)) {
-        ufsName = ufsName.substring(0, ufsName.length() - 1);
+    boolean synced = true;
+    if (!mUnchecked) {
```

Comment:
You can start with `if (mUnchecked) { return; }` to clean up the indentation a bit.